### ISSUE TITLE
Introduce a grace time for hostname/vip association

### ISSUE THREAD

**Issue by lahabana** [kumahq/kuma#2607]
### Summary

Currently if a service disappears and reappears in a short time there's no guarantee that it will maintain the same VIP.
It's also possible that a new service registering soon after the service disappearing will have the same IP.

This can cause issues:

- All services with 1 dp will change IP on every restart which will cause issues for anyone having any sort of DNS caching on the client-side
- In the case of DNS caching on the client if a service is replaced a client would try to connect to a different service rather than just failing to connect which would be a confusing behaviour

### Steps To Reproduce

1. Start 1 dp with serviceName `srv1`
2. Look at the IP of the service assigned to this DP `srv1.mesh` (on an empty cluster that will be 240.0.0.1)
3. Stop the dp
4. Start another dp with a different serviceName `srv2`
5. Start the previous dp `srv1`
6. See that `srv2.mesh --> 240.0.0.1` and `srv1.mesh --> 240.0.0.2` 

We would expect: `srv1.mesh` to remain at `240.0.0.1`

### Suggested change

TL;DR: Introduce a grace timeout to avoid reusing IPs for a while after the service stopped.

This involves:

- Adding a new configuration `kuma_dns_server_grace_time` with a default of 1h
- In `VirtualOutbound` add the `deletion_timestamp`
- In `VirtualOutboundMeshView.Update()` set the `deletion_timestamp` instead of removing the entry, remove entries that are passed their `deletion_timestamp` and the `deletion_timestamp` if an `OutboundEntry` is added to a virtualOutbound in grace.

### Open questions

1. Should we still resolve hostnames that are in grace?
2. I believe the reallocation shouldn't happen thanks to the way `GlobalView` works in the vips_allocator but we should check this belief.

**Comment by jpeach** [comment]
Do DNS records for VIPs have a TTL? Maybe the grace period should be some multiple of the TTL?

**Comment by lahabana** [comment]
Good question there's no TTL set but there might be a default one.

**Comment by lahabana** [comment]
Triage: There's a TTL in the Envoy config we should make this a multiple as James suggested

**Comment by lahabana** [comment]
@michaelbeaumont and @jakubdyszkiewicz what's the story for this in the new MeshService stuff?

**Comment by michaelbeaumont** [comment]
Well we essentially don't ever touch the VIPs so once the status is set, it's not going to be unassigned

**Comment by lahabana** [comment]
@michaelbeaumont so can we close this?

**Comment by michaelbeaumont** [comment]
I think so. If you delete the MeshService itself of course it's gone but that's fine
